source/extensions/outlier_detection_monitors/common/monitor_base_impl.h

```diff
@@ -0,0 +1,120 @@
+public:
+  HTTPCodesBucket() = delete;
+  HTTPCodesBucket(uint64_t start, uint64_t end) : start_(start), end_(end) {}
+  bool match(const ExtResult&) const override;
+
+  virtual ~HTTPCodesBucket() {}
+
+private:
+  uint64_t start_, end_;
+};
+
+// Class defines a "bucket" which catches LocalOriginEvent.
+class LocalOriginEventsBucket : public TypedErrorsBucket<Upstream::Outlier::LocalOriginEvent> {
+public:
+  LocalOriginEventsBucket() = default;
+  bool match(const ExtResult&) const override;
+};
+
+class ExtMonitorBase : public ExtMonitor {
+public:
+  ExtMonitorBase(const std::string& name, uint32_t enforce) : name_(name), enforce_(enforce) {}
+  ExtMonitorBase() = delete;
+  virtual ~ExtMonitorBase() {}
+  void putResult(const ExtResult&) override;
+
+  void setExtMonitorCallback(ExtMonitorCallback callback) override { callback_ = callback; }
+
+  void reset() override { onReset(); }
+  std::string name() const { return name_; }
+
+  void processBucketsConfig(
+      const envoy::extensions::outlier_detection_monitors::common::v3::ErrorBuckets& config);
+  void addErrorBucket(ErrorsBucketPtr&& bucket) { buckets_.push_back(std::move(bucket)); }
+
+protected:
+  virtual bool onError() PURE;
+  virtual void onSuccess() PURE;
+  virtual void onReset() PURE;
+  virtual std::string getFailedExtraInfo() { return ""; }
+
+  std::string name_;
+  uint32_t enforce_{100};
+  ExtMonitor::ExtMonitorCallback callback_;
+  std::vector<ErrorsBucketPtr> buckets_;
+};
+
+template <class ConfigProto>
+class ExtMonitorFactoryBase : public Upstream::Outlier::ExtMonitorFactory {
+public:
+  ExtMonitorCreateFn createMonitor(const std::string& monitor_name,
+                                   ProtobufTypes::MessagePtr&& config,
+                                   const ExtMonitorFactoryContext& context) override {
+    // This should throw exception if config is wrong. In the lambda below there is no need to
+    // validate config again.
+    Envoy::MessageUtil::downcastAndValidate<const ConfigProto&>(*config,
+                                                                context.messageValidationVisitor());
+    // "convert" unique pointer to shared one.
+    std::shared_ptr<Protobuf::Message> cfg(config.release());
+    return [&monitor_name, cfg, &context, this]() {
+      return createMonitorFromProtoTyped(monitor_name, dynamic_cast<ConfigProto&>(*cfg), context);
+    };
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return std::make_unique<ConfigProto>();
+  }
+
+  std::string name() const override { return name_; }
+
+  ExtMonitorFactoryBase(const std::string& name) : name_(name) {}
+
+private:
+  virtual ExtMonitorPtr createMonitorFromProtoTyped(const std::string& monitor_name,
+                                                    const ConfigProto& config,
+                                                    const ExtMonitorFactoryContext& context) PURE;
```

Comment:
Please take `createFilterFactoryFromProto` as an example and read its code.

Here is my suggestion to this code segment and I added some additional comment to it to tell why:
```
template <class ConfigProto>
class ExtMonitorFactoryBase : public Upstream::Outlier::ExtMonitorFactory {
public:
  // See here:
  // 1. Use `const Protobuf::Message&` as input proto config type.
  // 2. Non-const context in most case. This is a common practice, not a strong rule.
  ExtMonitorCreateFn createMonitor(const std::string& monitor_name, const Protobuf::Message& config,
                                   ExtMonitorFactoryContext& context) override {
    // See here:
    // 1. downcast the common proto config to typed config and do some validation (validate the proto
    // based on the PGV annotations)
    return createMonitorFromProtoTyped(monitor_name,
                                       Envoy::MessageUtil::downcastAndValidate<const ConfigProto&>(
                                           config, context.messageValidationVisitor()),
                                       context);
  }

  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
    return std::make_unique<ConfigProto>();
  }

  std::string name() const override { return name_; }

  ExtMonitorFactoryBase(const std::string& name) : name_(name) {}

private:
  // See here:
  // 1. The specific implementation need to load the typed proto to an config object (if it defined one)
  // in this method. And the implementation may do some hard-code validations when loading the config object.
  // Finally, the loaded config object will be captured in the lambda. This could ensure the lambda will
  // **never** throw an exception or failed at the runtime.
  virtual ExtMonitorCreateFn createMonitorFromProtoTyped(const std::string& monitor_name,
                                                         const ConfigProto& config,
                                                         ExtMonitorFactoryContext& context) PURE;

  const std::string name_;
};
```

Reply:
Got the idea. Thanks for explaining! So, basically, the idea is to operate as long as possible on generic type `ConfigProto& config` and convert it into specialized type in the last call to "specialized" factory. I agree, this is cleaner.